src/gz/bufread.rs

```diff
@@ -280,108 +298,213 @@ impl<R: BufRead + Write> Write for GzEncoder<R> {
 /// ```
 #[derive(Debug)]
 pub struct GzDecoder<R> {
-    inner: CrcReader<deflate::bufread::DeflateDecoder<R>>,
-    header: Option<io::Result<GzHeader>>,
-    finished: bool,
+    inner: GzState,
+    header: Option<GzHeader>,
+    reader: CrcReader<deflate::bufread::DeflateDecoder<R>>,
+    multi: bool
+}
+
+#[derive(Debug)]
+enum GzState {
+    Header(Vec<u8>),
+    Body,
+    Finished(usize, [u8; 8]),
+    Err(io::Error),
+    End
+}
+
+struct Buffer<'a, T> {
+    buf: io::Take<io::Cursor<&'a mut Vec<u8>>>,
+    reader: &'a mut T
+}
```

Comment:
I think this can probably be replaced with `Chain<&[u8], T>` perhaps? Or rather if you use `io::Read::chain` where the first part is the `&[u8]` slice and the second is `&mut T`, I think this adapter may not be needed?

Reply:
This adapter is different from `Chain`, which copies data into `buf` at read time so that it can be reused when an error occurs.

I think that a simple `Chain<&[u8], T>` can't do this, we still need a `Buffer<Chain<&[u8], T>>`.

Reply:
Ah ok, makes sense!